### PR TITLE
Bandwidth allocation pipeline data

### PR DESCRIPTION
**Pull request by aleitner**
This change is [Reviewable](https://reviewable.io/reviews/storj/storj/276)

Server
- [x] Retrieve
- [x] Store

Client
- [x] Retrieve
- [x] Store

### CODE REVIEW

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 1]
```diff
@@ -69,51 +71,86 @@ func (s *StreamWriter) Close() error {
...
+			}
+
+			if err = stream.Send(msg); err != nil {
+				sr.err = err
+				break
+			}
+
+			sr.allocated += nextTotal
+
+			if err = sr.throttle.ProduceAndWaitUntilBelow(int64(signer.bandwidthMsgSize), max); err != nil {
```
Comment (egonelbre) [inline pkg/piecestore/rpc/client/readerwriter.go]: produce exactly the same amount that was added into nextTotal

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 2]
```diff
@@ -69,51 +71,86 @@ func (s *StreamWriter) Close() error {
...
-
-		if err = stream.Send(msg); err != nil {
+		resp, err := stream.Recv()
+		if err != nil {
 			return nil, err
 		}
 
-		resp, err := stream.Recv()
+		sr.totalRead += int64(len(resp.GetContent()))
+		_, err = sr.throttle.ConsumeOrWait(int64(len(resp.GetContent())))
```
Comment (egonelbre) [inline pkg/piecestore/rpc/client/readerwriter.go]: don't wait here... if the other side is generous, there's no need to be picky

File: internal/sync2/throttle_test.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+	wg.Wait()
+
+	fmt.Println("done", throttle.Err())
+}
+
+func TestThrottleBasic(t *testing.T) {
+	throttle := NewThrottle()
+	var stage int64
+	c := make(chan error)
```
Comment (egonelbre) [inline internal/sync2/throttle_test.go]: add a capacity here

File: internal/sync2/throttle_test.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+	// consumer
+	go func() {
+		consume, _ := throttle.ConsumeOrWait(4)
+		if v := atomic.LoadInt64(&stage); v != 1 || consume != 4 {
+			c <- fmt.Errorf("did not block in time: %d / %d", v, consume)
```
Comment (egonelbre) [inline internal/sync2/throttle_test.go]: return here

File: internal/sync2/throttle_test.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		consume, _ = throttle.ConsumeOrWait(4)
+		if v := atomic.LoadInt64(&stage); v != 1 || consume != 4 {
+			c <- fmt.Errorf("did not block in time: %d / %d", v, consume)
```
Comment (egonelbre) [inline internal/sync2/throttle_test.go]: return here

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 6]
```diff
@@ -69,51 +71,93 @@ func (s *StreamWriter) Close() error {
...
+			}
+
+			if err = stream.Send(msg); err != nil {
+				sr.throttle.Fail(err)
+				return
+			}
+
+			sr.allocated += trustedSize
+
+			if err = sr.throttle.ProduceAndWaitUntilBelow(nextAllocSize, MaxAllocSize*2); err != nil {
```
Comment (egonelbre) [inline pkg/piecestore/rpc/client/readerwriter.go]: `int64(signer.bandwidthMsgSize*2)` or `*3`... the ideal would such number that the next allocation reaches the same time the storagenode runs out of allocations (with some margin of error)

File: internal/sync2/throttle_test.go (hunk 3 above)
Comment (aleitner) [reply internal/sync2/throttle_test.go]: What do you mean by capacity?

File: pkg/piecestore/rpc/client/readerwriter.go (hunk 6 above)
Comment (aleitner) [reply pkg/piecestore/rpc/client/readerwriter.go]: Are you referring to int64(signer.bandwidthMsgSize*2) instead of nextAllocSize?

Comment (egonelbre) [reply pkg/piecestore/rpc/client/readerwriter.go]: `sr.throttle.ProduceAndWaitUntilBelow(nextAllocSize, int64(signer.bandwidthMsgSize*2))`

Comment (egonelbre) [reply pkg/piecestore/rpc/client/readerwriter.go]: this assumes that the storage node is always sending back in chunks of `bandwidthMsgSize`

File: internal/sync2/throttle_test.go (hunk 3 above)
Comment (egonelbre) [reply internal/sync2/throttle_test.go]: `c := make(chan error, 1)` - this ensures that when both goroutine and the main fail, it won't get stuck.

File: pkg/piecestore/rpc/server/retrieve.go [hunk 7]
```diff
@@ -81,59 +85,89 @@ func (s *Server) retrieveData(ctx context.Context, stream pb.PieceStoreRoutes_Re
...
 
-		baData := &pb.RenterBandwidthAllocation_Data{}
+			// TODO: break when lastTotal >= allocData.GetPayer_allocation().GetData().GetMax_size()
 
-		if err = proto.Unmarshal(ba.GetData(), baData); err != nil {
-			return am.Used, am.TotalAllocated, err
-		}
+			if lastTotal > allocData.GetTotal() {
+				allocationTracking.Fail(fmt.Errorf("got lower allocation was %v got %v", lastTotal, allocData.GetTotal()))
+			}
```
Comment (egonelbre) [inline pkg/piecestore/rpc/server/retrieve.go]: return here

Comment (aleitner) [reply pkg/piecestore/rpc/server/retrieve.go]: Maybe we should continue instead?

File: internal/sync2/throttle.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
+package sync2
```
Comment (jtolds) [inline internal/sync2/throttle.go]: we should probably add a copyright header to this file

File: internal/sync2/throttle_test.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
+package sync2
```
Comment (jtolds) [inline internal/sync2/throttle_test.go]: same

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 10]
```diff
@@ -7,6 +7,8 @@ import (
 	"fmt"
 	"log"
 
+	"storj.io/storj/internal/sync2"
```
Comment (jtolds) [inline pkg/piecestore/rpc/client/readerwriter.go]: the imports here should be reordered.

```
stdlib
stdlib

thirdparty
thirdparty

storj
storj
```

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 11]
```diff
@@ -69,51 +71,94 @@ func (s *StreamWriter) Close() error {
...
-		updatedAllocation := int64(signer.bandwidthMsgSize) + sr.totalRead
-		allocationData := &pb.RenterBandwidthAllocation_Data{
-			PayerAllocation: pba,
-			Total:           updatedAllocation,
-		}
-
-		serializedAllocation, err := proto.Marshal(allocationData)
-		if err != nil {
-			return nil, err
+	trustLimit := int64(signer.bandwidthMsgSize * 64)
```
Comment (jtolds) [inline pkg/piecestore/rpc/client/readerwriter.go]: this should perhaps be a flag?

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 12]
```diff
@@ -69,51 +71,94 @@ func (s *StreamWriter) Close() error {
...
+	sendThreshold := int64(signer.bandwidthMsgSize * 8)
```
Comment (jtolds) [inline pkg/piecestore/rpc/client/readerwriter.go]: also this?

File: pkg/piecestore/rpc/client/readerwriter.go [hunk 13]
```diff
@@ -69,51 +71,94 @@ func (s *StreamWriter) Close() error {
...
+
+	// Send signed allocations to the piece store server
+	go func() {
+		trustedSize := int64(signer.bandwidthMsgSize * 8)
```
Comment (jtolds) [inline pkg/piecestore/rpc/client/readerwriter.go]: also this?

File: pkg/piecestore/rpc/server/retrieve.go [hunk 14]
```diff
@@ -81,59 +85,90 @@ func (s *Server) retrieveData(ctx context.Context, stream pb.PieceStoreRoutes_Re
...
+		if err != nil {
+			allocationTracking.Fail(err)
+			break
 		}
 
+		used += nextMessageSize
+		n, err := io.CopyN(writer, storeFile, nextMessageSize)
+		// correct errors when needed
+		if n != nextMessageSize {
+			if err := allocationTracking.Produce(nextMessageSize - n); err != nil {
```
Comment (jtolds) [inline pkg/piecestore/rpc/server/retrieve.go]: i'm nervous about future refactors here stomping on the 'err' variable with a scope change. maybe can we make the variable be different than err on this one line?

**Pull request merged**